### DISCUSSION TITLE
Avoiding quadratic stack-scanning behavior on very large stacks?

### DISCUSSION THREAD

**Discussion by gasche** [ocaml/ocaml#11923] [Ideas]
## Context

I was wondering whether TMC is actually still useful in the Multicore runtime, or if we shouget rid of it.

In the Multicore OCaml runtime, the runtime manages its own stack, which gives us freedom to write programs using a lot of stack space without running into operating-system limits. The Multicore runtime sets a stack-size limit by default and will raise Stack_Overflow exceptions when it is reached; this is a usability device to stop early on infinite stack-eating loops instead of consuming all the system memory and swapping. But the user can easily augment the stack limit with `OCAMLRUNPARAM="l=123456789", which used to work only in bytecode but now also works in native code. (The details of how Multicore uses its stack limit may change, but the above should remain true in the future.)

(In particular, we can now get rid of the threaded Stack_Overflow errors when compiling large/generated OCaml programs. But we are still at the mercy of quadratic/exponential slowdowns caused by algorithms.)

So what's stopping us from passing very large stack-limits if a program hits Stack_overflow due to (terminating) non-tail-recursive functions?

I did some expriments, and I observe a quadratic slowdown when the stack gets very large, due to repeated scanning of the whole stack by the minor collector. (See benchmark code at the end.) I'm opening the present issue to discuss the problem.

## Quadratic stack scanning

This phenomenon already existed with the sequential runtime, but it was rarely observed due to system stack limits. The sequential runtime has an optional optimization to mark the part of the stacks that were already scanned by a minor GC (so they cannot contain young roots anymore and don't need to be scanned again), but it was only enabled on some architectures:
- https://github.com/ocaml/ocaml/blob/4.13.0/runtime/roots_nat.c#L306-L311
- https://github.com/ocaml/ocaml/commit/f6a0392f57315eea2c9a750c534e40127d9940da
- https://github.com/ocaml/ocaml/commit/2395d35d86a59804bbabf0b1ea58b7994b07e3b5

What would be a good approach to optimize the Multicore runtime to avoid quadratic stack scanning, without overhead for programs that don't use a lot of stack? I see two potential strategies, using the fact that the Multicore runtime uses a linked list of contiguous stacks. (Normally, calls always go to the last contiguous stack, with resizing if necessary; only effect handlers create new stacks.)

1. We could ensure that the minor GC only scans the last/active stack, not previous stacks. (I tried to do this naively and my code doesn't work.) Then we can add a limit on contiguous stack size: when trying to grow the stack above that limit, we add a new stack instead. This bounds the amount of stack-scanning work per minor GC.

2. Or we could keep a "scanned up to that frame" information in the stack metadata, that would be updated "upward" (when pushing to the stack) by the stack-resizing code, and updated "downward" (when popping from the stack) by special call frames pushed by the stack-resizing code. We can decide how often to do those updates by tuning the resizing strategy (currently it is exponential, we would add an upward limit).

I wonder if that sounds reasonable to people, and something worth exploring.

(cc peopl with their name in relevant code files: @xavierleroy @stedolan @kayceesrk @ctk21)

## Measurement code

```ocaml
(* This non-recursive version sees a quadratic slowdown on very large
   lists, due to repeated scanning of the call stack by the minor
   GC. *)
let rec direct_map f xs =
  match xs with
  | [] -> []
  | x :: xs ->
      let y = f x in
      y :: direct_map f xs

(* The TMC version uses constant stack space; no slowdown. *)
let[@tail_mod_cons] rec tmc_map f xs =
  match xs with
  | [] -> []
  | x :: xs ->
      let y = f x in
      y :: tmc_map f xs

(* This version is similar to direct_map, except that every N calls it
   creates a new stack/fiber by using a dummy effect handler. If the
   minor GC would only scan the last fiber, this version should not
   suffer from a quadratic slowdown; but currently all stacks/fibers
   are scanned so this has no effect. *)
let handling_map f xs =
  let handler_threshold = 100_000 in
  let rec map f xs count =
    if count = 0 then
      map_on_new_stack f xs
    else
      match xs with
      | [] -> []
      | x :: xs ->
          let y = f x in
          y :: map f xs (count - 1)
  and map_on_new_stack f xs =
    let open Effect.Shallow in
    let continue () = start_mapping f xs in
    continue_with (fiber continue) () {
      retc = Fun.id;
      exnc = raise;
      effc = (fun _ -> None);
    }
  and start_mapping f xs =
    map f xs handler_threshold
  in start_mapping f xs 


let benchs = [
  "direct", direct_map;
  "handling", handling_map;
  "tmc", tmc_map;
]

let () =
  match
    List.assoc Sys.argv.(1) benchs,
    int_of_string Sys.argv.(2),
    int_of_string Sys.argv.(3) 
  with
  | exception _ ->
      Printf.eprintf "Usage: %s [%s] <list-size> <niter>\n%!"
        Sys.argv.(0)
        (String.concat "|" (List.map fst benchs))
  | map, list_size, niter ->
      let li = List.init list_size Fun.id in
      let old = ref [] in
      for i = 1 to niter do
        old := map succ li
      done
```

**Comment by xavierleroy** [comment]
Independently on stack size limits, I would expect TMC to result in lower memory usage, which in turn should also improve cache behavior.  It would be interesting to benchmark on well-chosen examples.

**Comment by xavierleroy** [comment]
Concerning possible optimizations for scanning roots in the stack:

That's something I looked into around 1996-1997, experimenting with some of the approaches you mention and hoping to write a PLDI submission out of it.  Measurements showed no significant speedups, and occasional slowdowns.  The only approach that was obviously always a win was to use the low bit of return addresses to mask already-scanned frames, on processors that ignore the low bit of the return address (i.e. Alpha and PowerPC).

So, I did not submit, but our CMU friends published essentially the same results at PLDI 1998, with more positive measurements because their baseline was less optimized than OCaml: https://doi.org/10.1145/277650.277718 .

It would be good to look at their paper before reinventing generational stack scanning.  Note, however, that the parts of the paper dealing with callee-save registers are irrelevant to OCaml because OCaml doesn't use callee-save registers.

**Comment by xavierleroy** [comment]
> The only approach that was obviously always a win was to use the low bit of return addresses to mask already-scanned frames, on processors that ignore the low bit of the return address (i.e. Alpha and PowerPC).

Looks like this trick should work on RISC-V too.  Hurrah!

**Comment by jhjourdan** [comment]
Perhaps that's naive, but how bad would it be to mask out the lower bit of the return address before using the `ret` instruction?

**Comment by xavierleroy** [comment]
> Perhaps that's naive, but how bad would it be to mask out the lower bit of the return address before using the ret instruction?

I recall that's one of the things I tried in 1996-1997 and it caused noticeable slowdowns on the mutator.  CPUs of the time did not do return address prediction, so any additional computation on the return address would delay the execution of the return instruction.  For x86 it was even worse because the return address is read from memory, so a read-modify-write instruction is inserted on the critical path to the return instruction.

Nowadays you can expect dynamic prediction of the return address, so maybe the impact is lower, but I would still be worried about increase in code size.

**Comment by gadmm** [comment]
The current test for stack capacity in function prologues could be replaced by a test of whether the top of stack falls inside safe windows placed at regular intervals from the stack limit. This can be tested efficiently with a `sub` and a `test` instruction with a statically-known mask. This happens to preserve the emitted code size (14B in total for the check in both cases on x86). Once you can do upwards checks (in @gasche's terminology) at regular intervals, say every 1 kB of stack, then you are able to determine the start of the minor roots within a margin of 1 kB using his strategy 2., without the need for the downwards checks that made it complicated to implement.

If this starts reminding you of Cheney on the MTA, another thing that could make CPS needless for avoiding high stack usage (on-topic for this issue) would be to please have memprof callbacks run for stack allocation. This lets you treat stack allocation like heap allocation in terms of memory limits, to avoid issues with the OOM killer that exists with high stack limits.

**Comment by kayceesrk** [comment]
All ARMv8 AArch64 processors have [top-byte-ignore](https://developer.arm.com/documentation/ka005620/latest/) feature enabled. This feature is necessary for ARM memory tagging extensions (MTE), which uses bits [59:56]. The bits [63:60] are available to be used for generational stack collection. [This article](https://www.linaro.org/blog/top-byte-ignore-for-fun-and-memory-savings/#:~:text=Where%20Can%20You%20Use%20TBI%3F) reports that both Linux and macOS support this feature.

Similar feature is also available on x86-64. AMD has [UAI](https://www.phoronix.com/news/AMD-Linux-UAI-Zen-4-Tagging) and Intel has [LAM](https://lpc.events/event/11/contributions/1010/attachments/875/1679/LAM-LPC-2021.pdf). However, it looks like LAM support will be [disabled in Linux due to security concerns](https://www.phoronix.com/news/Linux-Disabling-Intel-LAM).

**Comment by xavierleroy** [comment]
I learned about top-byte-ignore yesterday.  It is controlled by a bit in a special register that is not accessible from user mode.  From a quick experiment, Linux sets this bit by default, but macOS does not: I'm getting a segfault when jumping to a code pointer with bit 63 set.

**Comment by xavierleroy** [comment]
For the record: 
- POWER ignores the low 2 bits of function pointers
- RISC-V ignores the low bit of function pointers.

**Comment by stedolan** [comment]
> Similar feature is also available on x86-64. AMD has [UAI](https://www.phoronix.com/news/AMD-Linux-UAI-Zen-4-Tagging) and Intel has [LAM](https://lpc.events/event/11/contributions/1010/attachments/875/1679/LAM-LPC-2021.pdf). However, it looks like LAM support will be [disabled in Linux due to security concerns](https://www.phoronix.com/news/Linux-Disabling-Intel-LAM).

LAM won't help here: it applies only to data accesses, not code ones.

> LAM is supported only in 64-bit mode and applies only addresses used for data accesses. LAM doe not apply to
addresses used for instruction fetches or to those that specify the targets of jump and call instructions.

(from https://cdrdv2-public.intel.com/774990/architecture-instruction-set-extensions-programming-reference.pdf , pg. 127 / 6-1)

**Reply by lpw25** [reply]
Random idea that I've put essentially no thought into: when frame pointers are turned on could we rely on LAM for those to get some free bits?

**Reply by stedolan** [reply]
No, frame pointers are consumed by other tools (debuggers, perf, etc.) that do not necessarily respect LAM.

**Comment by kayceesrk** [comment]
One of the ideas that @stedolan mentioned is to expend an extra word per frame for keeping track of already-scanned information. Since we only need 1-bit of info, we could use the rest of the bits for something else. In particular, we could use the rest of the bits to store the offset to the frame table entry for this particular return address, avoiding the hash table lookup to find the frame descriptor entry from the return address. Given that we are no longer limited by the OS stack size limits, expending an extra word seems ok.

**Comment by gasche** [comment]
My (newcomer) thinking about this comes back to the idea of trying to pay book-keeping costs only in 1/Nth of the stack frames, which as @xavierleroy pointed out is basically the idea in the [PLDI'98 paper on generational stack scanning (and other optimizations)](https://dl.acm.org/doi/10.1145/277650.277718).

I think that we could reuse the logic of stack fragments that we already have to support algebraic effects:
- instead of storing one young-or-old word in each stack frame, store one yong-or-old word per stack *fragment*
- currently we have a stack check that the entry of stack-allocating functions that performs exponential resizing; tweak this logic to switch to a new stack fragment once the current stack fragment becomes larger than a certain limit (easier said than done? I'm not sure how to have a conditional check that sometimes behaves like a tail-call)

This scheme would have no code-size cost, and neglectible memory cost. The downside of course is that it is approximate, we still need to scan the current stack fragment in full.

**Reply by gadmm** [reply]
If you have "virtual" stack segments (e.g. by using the trick to avoid downwards checks I described above) you can avoid the boundary issue by resetting virtual segments such that the current frame falls in the middle, whenever the stack check logic is triggered (so that it has to move up or down the stack at least n KB before triggering the stack check again).

**Comment by xavierleroy** [comment]
The difficulty with segmented stacks is to avoid the situation where the mutator code repeatedly crosses the border between two fragments.  E.g. "f" is at the end of its fragment, calls "g", which needs a new fragment, then "g" returns to "f", switching back to the previous fragment (?), then "f" calls "g", etc.

**Comment by xavierleroy** [comment]
For a revival of the OCaml 4 approach, see #13594.

**Comment by NickBarnes** [comment]
I'm not sure what segmented stack approaches have been tried before in OCaml, but there's a variety of techniques for avoiding thrashing when you have the option of a hybrid approach.
For instance, we could use segments by default, but if we pop to the same segment several times within some short time horizon, or if we hit a stack check too soon after we last popped, we could take that as an indication to grow the current segment instead.
A related hack: initialize a stack segment with several markers, and use these when popping a stack segment to measure (approximately) how large the stack grew within it. If then one chooses to grow the parent segment, this can be a guide to a suitable growth size.

**Comment by xavierleroy** [comment]
> I'm not sure what segmented stack approaches have been tried before in OCaml

None, as far as I know.  I'm sure there are clever implementations in Scheme compilers (for incremental stack copying after call/cc) but I fear the complexity.
